### PR TITLE
[Util] Add a type interface for controlling hoisting into globals

### PR DESCRIPTION
**Pull request by qedawkins**
This adds a type interface that allows
a) Control over when to hoist a value of a particular type into a global
b) Inserting casting logic around the global to set preferred storage
types for the global

One implementation is provided for the ranked tensor type, with the idea
being, for non-byte aligned types, we want to avoid const-eval
generating such types directly as MLIR won't store them packed. This
allows us to keep data packed all the way through the compiler and never
exposes the runtime to a non-byte aligned type for this particular flow.
This will be especially helpful when moving to large arrays of structs.

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/Util/Analysis/Constant/OpOracle.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/Dialect/Util/Analysis/Constant/OpOracle.cpp:169]: This still needs a better way to convey some sort of contract w.r.t. what is valid to hoist, given that this PR is relying on bitcasting defined with a user hook rather than on the analysis itself.

File: compiler/src/iree/compiler/Dialect/Util/IR/UtilInterfaces.td
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Util/IR/UtilInterfaces.td]: set/unset is a bit hard to read - maybe more common terminology of encode/decode or something? (using `set` when you also have a `get` makes this look like some stateful parameter you are changing, not performing a transformation of a value)

**Comment by Max191** [comment]
+1
I was able to use this for dispatch tracing of `i4` tensors with `--iree-flow-trace-dispatch-tensors`. The hoisting interface is very useful for this.

File: compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp
Comment (benvanik) [inline compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp:1]: if you replace IREE::Flow::TensorBitCastOp with tensor::BitCastOp you can avoid this whole Interfaces/ directory and instead put it in UtilExternalModels.cpp - less code to maintain that way and the logic here is mostly about handling the tensor type instead of anything particular to global optimization

**Review by benvanik — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp
Comment (qedawkins) [reply compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp:1]: tensor::BitCastOp only allows changing the element type :(

Comment (qedawkins) [reply compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp:1]: I need to do `tensor<64xi4>` -> `tensor<32xi8>`

Comment (qedawkins) [reply compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp:1]: I don't mind looking for a different place for this, but I'd rather not try to convince upstream that `tensor.bitcast` should be able to bitcast to an arbitrary tensor, at least not right now.

Comment (qedawkins) [reply compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp:1]: Also, it is sort of specific to GlobalOptimization. The `decodeStorageType` inserts a `flow.tensor.bitcast` (or whatever) which is fine before conversion to Stream, but if there was ever a tensor at a later stage, we might need a different op to bitcast back. At least that's been my thinking so far; if it's never legal to have a tensor that we can't use flow (or tensor) ops on, then finding a better place to register it makes sense.

**Comment by bjacob** [comment]
FYI @qedawkins we're trying this on @Max191 's branch on the turbine.py llama2 model and running into a problem that seems like it might be caused by the "b) Inserting casting logic around the global to set preferred storage types for the global" aspect of this PR ?

https://gist.github.com/bjacob/4c4eb81c3a85f8baa3c1c3cfbce79873

**Comment by qedawkins** [comment]
> FYI @qedawkins we're trying this on @Max191 's branch on the turbine.py llama2 model and running into a problem that seems like it might be caused by the "b) Inserting casting logic around the global to set preferred storage types for the global" aspect of this PR ?
> 
> https://gist.github.com/bjacob/4c4eb81c3a85f8baa3c1c3cfbce79873

Hmmm, I can take a look. The way this is structured shouldn't need to care about shapes, so something strange is happening.

File: compiler/src/iree/compiler/Dialect/Util/Transforms/HoistIntoGlobals.cpp
Comment (bjacob) [inline compiler/src/iree/compiler/Dialect/Util/Transforms/HoistIntoGlobals.cpp:125]: I wonder if this is what causes my issue?  The preexisting `originalValue.replaceAllUsesWith(load);` was perhaps assuming that the new shape equals the old shape, and that is no longer true? @qedawkins

Comment (bjacob) [reply compiler/src/iree/compiler/Dialect/Util/Transforms/HoistIntoGlobals.cpp:125]: I meant for this comment to be against the block of lines from 125 to 139, not just this line 125.

Comment (bjacob) [reply compiler/src/iree/compiler/Dialect/Util/Transforms/HoistIntoGlobals.cpp:125]: Ah no, nevermind, `originalValue` and `load` always have the same type in my testcase, the right one (`4096x4096xi4`)

Comment (qedawkins) [reply compiler/src/iree/compiler/Dialect/Util/Transforms/HoistIntoGlobals.cpp:125]: That's what the check on line 132 is supposed to be verifying, so maybe? It's probably something along those lines though.

**Comment by stellaraccident** [comment]
Just processing my notification queue. Is this PR ready to review for real or are comments still being addressed?

**Comment by benvanik** [comment]
(I missed the comment responses due to holidays, going to take another look now)

File: compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp
Comment (benvanik) [reply compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp:1]: tensor.bitcast being annoying strikes again! upsetting we can't have basic things upstream :(

I'm still very sketched out by adding external models that have tight coupling to our own dialects that have staging requirements. If this interface was used in GlobalOptimization only _and_ the HoistIntoGlobals pass was only used there it's fine as we know flow ops are valid to insert, but HoistIntoGlobals can run from anywhere at any time and may or may not want to insert the op - like canonicalization patterns there's no way to control this and HoistIntoGlobals running outside of GlobalOptimization inherits this behavior with no way to disable it. How things are used today may be fine with this but we want to hoist a lot more - including inside executables - and this behavior would block that (if we ran HoistIntoGlobals inside an executable still in linalg-on-tensors it'd create flow ops, which is not ok).

If this were on our own tensor type things get simpler as if you are using some `!flow.tensor` you'd expect to get `flow.tensor.bitcast` ops - and if you weren't using the flow type (just `tensor`) then you wouldn't get flow ops. Making flow use its own type has been something I've played with but it's hard due to incremental dispatch region formation - we don't want to have `tensor` and some `flow.tensor` having to exist at the same time and be handled by everywhere operating on tensor-like things. If there were a TensorTypeInterface that the upstream TensorType used such that we could extend things that'd be great, but AFAIK it's still all locked-down.

The solution is to have an upstream tensor op that can do an actual friggin bitcast. If we have to do this for now then we have to... am open to suggestions on how to avoid the coupling - maybe an `util.whatever_just_cast_you_stupid_compiler` op or something... but a `tensor.bitcast_in_the_truest_meaning_of_the_word` is really what we want. Any ideas?

Comment (qedawkins) [reply compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp:1]: > If this interface was used in GlobalOptimization only and the HoistIntoGlobals pass was only used there it's fine as we know flow ops are valid to insert, but HoistIntoGlobals can run from anywhere at any time and may or may not want to insert the op - like canonicalization patterns there's no way to control this and HoistIntoGlobals running outside of GlobalOptimization inherits this behavior with no way to disable it. How things are used today may be fine with this but we want to hoist a lot more - including inside executables - and this behavior would block that (if we ran HoistIntoGlobals inside an executable still in linalg-on-tensors it'd create flow ops, which is not ok).

Agree, not ok to create flow ops inside an executable, hence why I registered the interface in GlobalOptimization only (although I assume this interface can't be "un-registered," so I'm probably not actually solving anything :P ). The important thing is that if we have an interface registered globally and consumed by HoistIntoGlobals, every consumer of HoistIntoGlobals then needs to be able to handle the bitcasts that come out of it.

> The solution is to have an upstream tensor op that can do an actual friggin bitcast. If we have to do this for now then we have to... am open to suggestions on how to avoid the coupling - maybe an util.whatever_just_cast_you_stupid_compiler op or something... but a tensor.bitcast_in_the_truest_meaning_of_the_word is really what we want. Any ideas?

I would have used it if it was there. `util.give_me_my_tensor<123xi4>` works for me, and then each consumer can lower the op to whatever they need themselves. The main concern in codegen would be the need to use upstream interfaces (e.g. bufferization, xyz) to handle the op, and I assume that kind of stuff wouldn't be something we'd want on a util op, bringing us back to `tensor.bitcast_for_real_hacked_twin`.

I don't have the bandwidth to push on `tensor.bitcast` upstream right now, and it would be good to unblock the CPU work that is relying on this. What about

1) In this PR, or as a separate PR, we add `util.bitcast` that has a lowering to `flow.bitcast` or we just lower it directly to stream. This makes the interface registration slightly more sane, at the expense of adding an additional util op.
2) In parallel, someone starts an RFC for tensor.bitcast, essentially aiming to be a drop in replacement for the util version once available.

Comment (stellaraccident) [reply compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp:1]: What am I missing: We have a `tensor.bitcast` (https://github.com/llvm/llvm-project/blame/main/mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td#L60). What is it doing wrong?

Comment (stellaraccident) [reply compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp:1]: ah, the whole "equivalent element width" thing. I think we should just remove that constraint and use language that is more about this being legal if the lowering can resolve it to a bag of bytes of the same volume or something.

Comment (qedawkins) [reply compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp:1]: Yeah. Maybe a good middle ground is to push for statically shaped bitcasts only for now.

Comment (qedawkins) [reply compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp:1]: Maybe that's what you meant by
```
this being legal if the lowering can resolve it to a bag of bytes of the same volume or something.
```

Comment (stellaraccident) [reply compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp:1]: Ok, nevermind - yeah, this is more of an element_type bitcast op. I think we should just add an op like `reinterpret_cast` which is defined to be contextually legal in lowering flows that resolve the whole tensor to a bag of bytes of equivalent size.

Casts are hard.

Comment (qedawkins) [reply compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp:1]: If this is a blocker for this PR, I could use some help pushing that upstream. I have a lot of other things on my plate right now and don't have room to add another upstream discussion to it.

Comment (qedawkins) [reply compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp:1]: (if it's going the util route though, I don't mind doing that as a part of this PR).

Comment (stellaraccident) [reply compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp:1]: If we can untangle it later, I'll defer to Ben. Can add an upstreaming chore.

Comment (benvanik) [reply compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp:1]: yeah, registration is global and happens on startup, like canonicalization patterns. it at least keeps it scoped in the compiler if we ever wanted to eject parts (global optimization moves to a plugin or something) but doesn't solve the problem.

I think we land this now and then ideally get some help on fixing upstream - it's a functionality gap to not have a proper tensor-compatible bitcast and is something nearly anyone working on interop layers needs (we achieve it via our hal.tensor.import/export today, for example).

thanks for brainstorming on this :) after the canonicalization pattern revert upstream the other week for similar reasons this kind of issue is top of mind for me!

**Review by benvanik — APPROVED** [review]

File: compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp
Comment (benvanik) [reply compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp:1]: (whoops that whole conversation happened while I was typing that - re-reading)

Comment (benvanik) [reply compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp:1]: yeah, tensor.reinterpret_cast is really what we want - I think tensor.bitcast got pushback because some random pass someone had needed to be able to look through it with affine maps or something, but we really don't care about that and if a new op that doesn't allow that kind of analysis existed I suspect there'd be less worry.... but I'm bad at predicting upstream ;)

Comment (qedawkins) [reply compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp:1]: Sounds good! I'm going to wait on @bjacob 's approval too because it seems like there's a chance of a bug.

Comment (bjacob) [reply compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp:1]: As discussed offline with @Max191 and @qedawkins :

Testcase: <https://gist.github.com/bjacob/c90cfb319fc87296636b4111c9c1c056>
Repro: `iree-compile --iree-opt-const-eval=false  --iree-hal-target-backends=llvm-cpu --iree-global-opt-enable-quantized-matmul-reassociation --iree-opt-data-tiling`

Succeeds at this PR's current branch, but fails once we rebase this PR on current `main` branch. When it fails, the specific commit in this PR that is the first to fail is the first one: https://github.com/openxla/iree/pull/15558/commits/285f5835bd4798b26f7cf852c4b857b6e62b8d41

Comment (bjacob) [reply compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp:1]: now bisecting where on `main` branch the change happened that makes rebasing this PR there, reproduce.

Comment (bjacob) [reply compiler/src/iree/compiler/GlobalOptimization/Interfaces/HoistableTypeInterface.cpp:1]: The first commit on `main` branch where rebasing this PR's first commit https://github.com/openxla/iree/commit/285f5835bd4798b26f7cf852c4b857b6e62b8d41 allows to reproduce, is: https://github.com/openxla/iree/commit/1f8d0acc987956c5186329e889ea18b9b0d438fc

**Comment by qedawkins** [comment]
After some offline discussion with @bjacob I'm going to land this now because the problem this is hitting related to this is not in main yet and it helps to not have to keep rebasing this PR.

**Pull request merged**
